nixos/modules/services/web-apps/lemmy.nix

```diff
@@ -0,0 +1,142 @@
+{ lib, pkgs, config, ... }:
+with lib;
+let
+  cfg = config.services.lemmy;
+  settingsFormat = pkgs.formats.json { };
+in
+{
+  options.services.lemmy = {
+    enable = mkEnableOption "Enable lemmy server";
+
+    config = mkOption {
+      type = types.submodule {
+        freeformType = settingsFormat.type;
+        options.database = {
+          password = mkOption {
+            type = types.str;
+            description = "password to connect to postgres";
+          };
+        };
+        options.hostname = mkOption {
+          type = types.str;
+          default = null;
+          description = ''the domain name of your instance (eg "lemmy.ml")'';
+        };
+        options.port = mkOption {
+          type = types.port;
+          default = 8536;
+          description = "port where lemmy should listen for incoming requests";
+        };
+        # TODO: this is a secret, should not be in nix-store
+        options.jwt_secret = mkOption {
+          type = types.str;
+          description = "json web token for authorization between server and client";
+        };
+        options.federation = {
+          enabled = mkEnableOption "whether to enable activitypub federation.";
+        };
+        options.captcha = {
+          enabled = mkOption {
+            type = types.bool;
+            default = true;
+            description = "Enable Captcha";
+          };
+          difficulty = mkOption {
+            type = types.str; # TODO: Types should be "easy", "medium", "hard". Define an enum.
+            default = "medium";
+            description = "The difficultly of the captcha to solve";
+          };
+        };
+      };
+      default = { };
+      description = "Lemmy configuration";
+    };
+
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.lemmy.settings = {
+      database.user = mkDefault "lemmy";
+      database.host = mkDefault "localhost";
+      database.port = mkDefault 5432;
+      database.database = mkDefault "lemmy";
+      database.pool_size = mkDefault 5;
+
+      bind = mkDefault "127.0.0.1";
+      tls_enabled = mkDefault true;
+      pictrs_url = mkDefault "http://localhost:8080";
+      actor_name_max_length = mkDefault 20;
+
+      rate_limit.message = mkDefault 180;
+      rate_limit.message_per_second = mkDefault 60;
+      rate_limit.post = mkDefault 6;
+      rate_limit.post_per_second = mkDefault 600;
+      rate_limit.register = mkDefault 3;
+      rate_limit.register_per_second = mkDefault 3600;
+      rate_limit.image = mkDefault 6;
+      rate_limit.image_per_second = mkDefault 3600;
+    };
+
+    # TODO: Check if this can't be put under /var/lib
+    environment.etc."lemmy.json".source =
+      settingsFormat.generate "config.hjson" cfg.settings;
+
+    services.postgresql.enable = true;
```

Comment:
- Does this software work with MariaDB as well?
- What if the database runs on a different server?

Reply:
It only works with postgres, but as you correctly pointed out, the database could be on another server.
I'm guessing that here we need to have perhaps two more options.
- `useLocalPostgres` to confirm that the user wants to use the local postgres
- `initialisePostgres` to confirm that the user wants to initialize the database with the given script and not do it himself.

Reply:
Made the change, feel free to comment again if you think this could use some improvements